rasterio/features.py

```diff
@@ -345,27 +361,43 @@ def rasterize(
             else:
                 raise ValueError("GDAL versions < 3.6 cannot rasterize int64 values.")
 
-    if out is not None:
-        pass
-
-    elif out_shape is not None:
-        if len(out_shape) != 2:
-            raise ValueError('Invalid out_shape, must be 2D')
-        out = np.empty(out_shape, dtype=dtype)
-        out.fill(fill)
+    with ExitStack() as exit_stack:
+        if dst_path is not None:
+            if isinstance(dst_path, DatasetWriter):
```

Comment:
Thoughts about removing `dst_path` and just using `out` instead?

Reply:
@snowman2 I've added `dst_path` and `dst_kwds` for harmony with the `merge()` and `stack()` tools and to partially satisfy part of the interface in `rasterio.tools`. Also, I wouldn't want to overload `out` so much, I think it's better if that's reserved for a array.